RoaringBitmap/src/main/java/org/roaringbitmap/art/Node.java

```diff
@@ -0,0 +1,283 @@
+   *
+   * @param current current internal node
+   */
+  public static Node insertLeaf(Node current, LeafNode childNode, byte key) {
+    switch (current.nodeType) {
+      case NODE4:
+        return Node4.insert(current, childNode, key);
+      case NODE16:
+        return Node16.insert(current, childNode, key);
+      case NODE48:
+        return Node48.insert(current, childNode, key);
+      case NODE256:
+        return Node256.insert(current, childNode, key);
+      default:
+        throw new IllegalArgumentException("Not supported node type!");
+    }
+  }
+
+  /**
+   * copy the prefix between two nodes
+   */
+  public static void copyPrefix(Node src, Node dst) {
+    dst.prefixLength = src.prefixLength;
+    System.arraycopy(src.prefix, 0, dst.prefix, 0, src.prefixLength);
+  }
+
+  public static void insertionSortOnNode4Or16(Node node) {
+    if (node.nodeType == NodeType.NODE4) {
+      Node4 node4 = (Node4) node;
+      sortSmallByteArray(node4.key, node4.children, 0, node4.count - 1);
+      return;
+    }
+    if (node.nodeType == NodeType.NODE16) {
+      Node16 node16 = (Node16) node;
+      sortSmallByteArray(node16.key, node16.children, 0, node16.count - 1);
+      return;
+    }
+    throw new IllegalArgumentException();
+  }
+
+  /**
+   * sort the small arrays through the insertion sort alg.
+   */
+  private static void sortSmallByteArray(byte[] key, Node[] children, int left, int right) {
+    for (int i = left, j = i; i < right; j = ++i) {
+      byte ai = key[i + 1];
+      Node child = children[i + 1];
+      int unsignedByteAi = Byte.toUnsignedInt(ai);
+      while (unsignedByteAi < Byte.toUnsignedInt(key[j])) {
+        key[j + 1] = key[j];
+        children[j + 1] = children[j];
+        if (j-- == left) {
+          break;
+        }
+      }
+      key[j + 1] = ai;
+      children[j + 1] = child;
+    }
+  }
+
+  public static int binarySearch(Node node, byte key) {
+    if (node.nodeType == NodeType.NODE4) {
+      Node4 node4 = (Node4) node;
+      return binarySearch(node4.key, 0, node4.count, key);
+    }
+    if (node.nodeType == NodeType.NODE16) {
+      Node16 node16 = (Node16) node;
+      return binarySearch(node16.key, 0, node16.count, key);
+    }
+    throw new IllegalArgumentException("Unsupported node type!");
+  }
+
+  private static int binarySearch(byte[] key, int fromIndex, int toIndex,
+      byte k) {
+    int inputUnsignedByte = Byte.toUnsignedInt(k);
+    int low = fromIndex;
+    int high = toIndex - 1;
+
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      int midVal = Byte.toUnsignedInt(key[mid]);
+
+      if (midVal < inputUnsignedByte) {
+        low = mid + 1;
+      } else if (midVal > inputUnsignedByte) {
+        high = mid - 1;
+      } else {
+        return mid; // key found
+      }
+    }
+    // key not found.
+    return ILLEGAL_IDX;
+  }
+
+  private void serializeHeader(DataOutput dataOutput) throws IOException {
```

Comment:
The preferred API for this since 0.8.0 serialises to a `ByteBuffer` because `DataOutput`/`DataInput` present lots of bottlenecks.

Reply:
agree, I left that as TODO to make this PR come out soon. I will implement that at a later time.